### PR TITLE
Partial evaluation support for assignment expressions and multiple array related expressions

### PR DESCRIPTION
**Pull request by cesarzc**
This change adds support in partial evaluation for the following expressions:
- Assignment.
- Array repeat.
- Array in-place concatenation.
- Array in-place update of value at index.
- Array access to value at index.

### CODE REVIEW

File: compiler/qsc_partial_eval/src/lib.rs [hunk 1]
```diff
@@ -230,23 +282,146 @@ impl<'a> PartialEvaluator<'a> {
...
+        // This is a workaround since evaluating sections of the execution graph that correspond to local variable
+        // expressions does not yield the expected results.
+        let expr = self.get_expr(expr_id);
+        if let ExprKind::Var(Res::Local(local_var_id), _) = &expr.kind {
+            let value = self
+                .eval_context
+                .get_current_scope()
+                .get_classical_local_value(*local_var_id);
+            return Ok(EvalControlFlow::Continue(value.clone()));
+        }
```
Comment (idavis) [inline compiler/qsc_partial_eval/src/lib.rs:406]: We should talk about this on Monday.

File: compiler/qsc_partial_eval/src/lib.rs [hunk 2]
```diff
@@ -370,32 +526,111 @@ impl<'a> PartialEvaluator<'a> {
...
+            return Err(Error::Unexpected(
+                "embedded return in array".to_string(),
+                value_expr.span,
+            ));
+        };
+        let size_control_flow = self.try_eval_expr(size_expr_id)?;
+        let EvalControlFlow::Continue(size) = size_control_flow else {
+            let size_expr = self.get_expr(size_expr_id);
+            return Err(Error::Unexpected(
+                "embedded return in array".to_string(),
```
Comment (idavis) [inline compiler/qsc_partial_eval/src/lib.rs:548]:
```suggestion
                "embedded return in array size".to_string(),
```

File: compiler/qsc_partial_eval/src/lib.rs [hunk 3]
```diff
@@ -1107,12 +1380,117 @@ impl<'a> PartialEvaluator<'a> {
     }
 
     fn try_eval_stmt(&mut self, stmt_id: StmtId) -> Result<EvalControlFlow, Error> {
-        // If the statement is classical, we can just evaluate it.
-        if self.is_classical_stmt(stmt_id) {
-            self.eval_classical_stmt(stmt_id)
+        let stmt = self.get_stmt(stmt_id);
+        match stmt.kind {
+            StmtKind::Expr(expr_id) => self.try_eval_expr(expr_id),
```
Comment (idavis) [inline compiler/qsc_partial_eval/src/lib.rs:1385]: Why doesn't handling of `StmtKind::Expr` have to handle control flow adjustments? Please add a comment.

**Review by idavis — APPROVED** [review]

File: compiler/qsc_partial_eval/src/lib.rs (hunk 1 above)
Comment (swernli) [reply compiler/qsc_partial_eval/src/lib.rs:406]: Is there a test case associated with this? I commented it out and everything still passes, so maybe it's not needed (or there is a test missing)?

File: compiler/qsc_partial_eval/src/lib.rs (hunk 3 above)
Comment (cesarzc) [reply compiler/qsc_partial_eval/src/lib.rs]: Comment added.

File: compiler/qsc_partial_eval/src/lib.rs (hunk 1 above)
Comment (cesarzc) [reply compiler/qsc_partial_eval/src/lib.rs]: Added some more test cases that without the fix or my initial workaround should fail (now they don't 😄)

**Review by swernli — APPROVED** [review]

**Pull request merged**